sql: handle zone config in ALTER PRIMARY KEY for REGIONAL BY ROW

otan opened the pull request:
Release note (sql change): REGIONAL BY ROW tables now preserve zone
configurations when using ALTER PRIMARY KEY.

Resolves #58339

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/59365)
<!-- Reviewable:end -->

ajstorm reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59365)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan)

___
*[pkg/ccl/logictestccl/testdata/logic_test/regional_by_row, line 201 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59365#-MRtwmMI2_gpxH0veGyV:-MRtwmMI2_gpxH0veGyW:bdrue3p) ([raw file](https://github.com/cockroachdb/cockroach/blob/aa2aaa8e3913ff8e2899483d48b8f9f92668b1bd/pkg/ccl/logictestccl/testdata/logic_test/regional_by_row#L201)):*
> ```
> ca-central-1    7   7   8   9
> ca-central-1    10  10  11  12
> us-east-1       20  20  21  22
> ```

Can we also create a test where `pk2` has duplicates to validate that the ALTER will fail?
___
*[pkg/ccl/logictestccl/testdata/logic_test/regional_by_row, line 302 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59365#-MRtwDVmB-sSjis_xkHt:-MRtwDVmB-sSjis_xkHu:bnhxy57) ([raw file](https://github.com/cockroachdb/cockroach/blob/aa2aaa8e3913ff8e2899483d48b8f9f92668b1bd/pkg/ccl/logictestccl/testdata/logic_test/regional_by_row#L302)):*
> ```
>   crdb_region public.crdb_internal_region NOT NULL DEFAULT gateway_region()::public.crdb_internal_region,
>   CONSTRAINT "primary" PRIMARY KEY (pk2 ASC),
>   UNIQUE INDEX regional_by_row_table_pk_key (pk ASC),
> ```

Is this correct?  Shouldn't this unique index be dropped as part of the ALTER?  Or is that still WIP?
___
*[pkg/sql/schema_changer.go, line 1422 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59365#-MRu-LjM5OxZA5z3DRr2:-MRu-LjM5OxZA5z3DRr3:b-phdpjy) ([raw file](https://github.com/cockroachdb/cockroach/blob/aa2aaa8e3913ff8e2899483d48b8f9f92668b1bd/pkg/sql/schema_changer.go#L1422)):*
> ```Go
> 	}
> 
> 	// It is safe to copy the zone config off a primary index of a REGIONAL BY ROW table.
> ```

Could we add a comment here as to "why"?  At first glance, I don't see why this is safe only for `REGIONAL BY ROW` tables.  Is it because other table locality patterns won't have sub-zones defined for their indexes?


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59365)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajstorm and @otan)

___
*[pkg/ccl/logictestccl/testdata/logic_test/regional_by_row, line 201 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59365#-MRtwmMI2_gpxH0veGyV:-MS-CnA-7YMlOGpDJbZi:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/aa2aaa8e3913ff8e2899483d48b8f9f92668b1bd/pkg/ccl/logictestccl/testdata/logic_test/regional_by_row#L201)):*
<details><summary><i>Previously, ajstorm wrote…</i></summary><blockquote>

Can we also create a test where `pk2` has duplicates to validate that the ALTER will fail?
</blockquote></details>

Done.
___
*[pkg/ccl/logictestccl/testdata/logic_test/regional_by_row, line 302 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59365#-MRtwDVmB-sSjis_xkHt:-MS-9smuAu4iMYK-y3jC:bvktryu) ([raw file](https://github.com/cockroachdb/cockroach/blob/aa2aaa8e3913ff8e2899483d48b8f9f92668b1bd/pkg/ccl/logictestccl/testdata/logic_test/regional_by_row#L302)):*
<details><summary><i>Previously, ajstorm wrote…</i></summary><blockquote>

Is this correct?  Shouldn't this unique index be dropped as part of the ALTER?  Or is that still WIP?
</blockquote></details>

This is existing behaviour of ALTER PRIMARY KEY -- it gets converted to a UNIQUE index and gets marked as being deleted.
We don't actually drop the index (yet) -- this is done asynchronously via some cleanup job but apparently logic tests don't trigger them.


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59365)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajstorm)

___
*[pkg/sql/schema_changer.go, line 1422 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59365#-MRu-LjM5OxZA5z3DRr2:-MS4Dy6_9TahXo3eL3qf:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/aa2aaa8e3913ff8e2899483d48b8f9f92668b1bd/pkg/sql/schema_changer.go#L1422)):*
<details><summary><i>Previously, ajstorm wrote…</i></summary><blockquote>

Could we add a comment here as to "why"?  At first glance, I don't see why this is safe only for `REGIONAL BY ROW` tables.  Is it because other table locality patterns won't have sub-zones defined for their indexes?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

ajstorm reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

Just nits and and one clarifying question.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59365#-:-MS5Px3aEz1uvSc7D1rZ:bxoj1x1)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajstorm and @otan)

___
*[pkg/ccl/logictestccl/testdata/logic_test/regional_by_row, line 302 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59365#-MRtwDVmB-sSjis_xkHt:-MS5LyaU9getfVHsB5qs:belmnxo) ([raw file](https://github.com/cockroachdb/cockroach/blob/aa2aaa8e3913ff8e2899483d48b8f9f92668b1bd/pkg/ccl/logictestccl/testdata/logic_test/regional_by_row#L302)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

This is existing behaviour of ALTER PRIMARY KEY -- it gets converted to a UNIQUE index and gets marked as being deleted.
We don't actually drop the index (yet) -- this is done asynchronously via some cleanup job but apparently logic tests don't trigger them.
</blockquote></details>

Hmm, also interesting.  Can we make a note of this in #58977  so that we ensure we test this in some way outside of testlogic?
___
*[pkg/ccl/logictestccl/testdata/logic_test/regional_by_row, line 313 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/59365#-MS5LhbHBDqia5M5XIZa:-MS5LhbHBDqia5M5XIZb:b-6uoyzh) ([raw file](https://github.com/cockroachdb/cockroach/blob/95f12e2cc87218d27a43c2e4ec5e884adf210c2e/pkg/ccl/logictestccl/testdata/logic_test/regional_by_row#L313)):*
> ```
> INSERT INTO regional_by_row_table (pk, pk2, a, b) VALUES (1000, 1, 1000, 2000)
> 
> statement error Key \(crdb_region,pk2\)=\('ap-southeast-2',1\) already exists.
> ```

Very interesting.  Does it make sense to hide the implicit part of this duplicate key?
___
*[pkg/sql/schema_changer.go, line 1412 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/59365#-MS5PJE71kST5GAmuB7C:-MS5PJE71kST5GAmuB7D:b1gt4i9) ([raw file](https://github.com/cockroachdb/cockroach/blob/95f12e2cc87218d27a43c2e4ec5e884adf210c2e/pkg/sql/schema_changer.go#L1412)):*
> ```Go
> 
> 	// It is safe to copy the zone config off a primary index of a REGIONAL BY ROW table.
> 	// This is because the prefix of the PK we used to partition by will stays the same,
> ```

Nit: "...will stay the same"

I'm also not following why we need this code down here.  Doesn't the loop above handle all indexes?  Is the PK index not in swapInfo?


<!-- Sent from Reviewable.io -->

ajstorm reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

This time with feeling.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59365#-:-MS5Q2uHC3bkIamMQis0:bhdy0a2)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajstorm and @otan)



<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59365)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajstorm)

___
*[pkg/ccl/logictestccl/testdata/logic_test/regional_by_row, line 302 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59365#-MRtwDVmB-sSjis_xkHt:-MS5T6FR7JLm8EcbCoaC:b-ndbdn9) ([raw file](https://github.com/cockroachdb/cockroach/blob/aa2aaa8e3913ff8e2899483d48b8f9f92668b1bd/pkg/ccl/logictestccl/testdata/logic_test/regional_by_row#L302)):*
<details><summary><i>Previously, ajstorm wrote…</i></summary><blockquote>

Hmm, also interesting.  Can we make a note of this in #58977  so that we ensure we test this in some way outside of testlogic?
</blockquote></details>

ah i lied, we allow this right now: https://github.com/cockroachdb/cockroach/blob/36f181a6fc0e695d48042fab5bcb08b70b61998e/pkg/sql/alter_primary_key.go#L374
if you feel like we should be rid of it, we can change it in a separate PR.
___
*[pkg/ccl/logictestccl/testdata/logic_test/regional_by_row, line 313 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/59365#-MS5LhbHBDqia5M5XIZa:-MS5S2BWAf1yktICmGm_:b-ft49c6) ([raw file](https://github.com/cockroachdb/cockroach/blob/95f12e2cc87218d27a43c2e4ec5e884adf210c2e/pkg/ccl/logictestccl/testdata/logic_test/regional_by_row#L313)):*
<details><summary><i>Previously, ajstorm wrote…</i></summary><blockquote>

Very interesting.  Does it make sense to hide the implicit part of this duplicate key?
</blockquote></details>

Yes, filed an issue.
___
*[pkg/sql/schema_changer.go, line 1412 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/59365#-MS5PJE71kST5GAmuB7C:-MS5TNtm9ZN8RZ1N_LkC:b-m52qq6) ([raw file](https://github.com/cockroachdb/cockroach/blob/95f12e2cc87218d27a43c2e4ec5e884adf210c2e/pkg/sql/schema_changer.go#L1412)):*
<details><summary><i>Previously, ajstorm wrote…</i></summary><blockquote>

Nit: "...will stay the same"

I'm also not following why we need this code down here.  Doesn't the loop above handle all indexes?  Is the PK index not in swapInfo?

</blockquote></details>

PK is kept as a separate field compared to the Index fields.


<!-- Sent from Reviewable.io -->

otan commented:
bors r=ajstorm

otan committed:
sql: handle zone config in ALTER PRIMARY KEY for REGIONAL BY ROW

Release note (sql change): REGIONAL BY ROW tables now preserve zone
configurations when using ALTER PRIMARY KEY.

otan commented:
bors r=ajstorm
